File: MyLeetCode/Leetcode/Guess Number Higher or Lower II.cpp
/*
amount[i][j]:money that need to guarantee a win in range(i,j)
set a l and h boundary and pick every number between l and h to calculate the money needed
Time complexity:O(n^2)
*/
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
class Solution {
public:
    int getMoneyAmount(int n) {
        vector<vector<int>> amount(n+1, vector<int>(n+1));
        for (int h = 2; h <= n; h++) {
            for (int l = h-1; l > 0; l--) {
                int globalmin = INT_MAX;
                for (int k = l; k <= h; k++) {
                    int local;
                    if (k == l) {
                        local = k + amount[k+1][h];
                    }
                    else if (k == h) {
                        local = k + amount[l][k-1];
                    }
                    else {
                        local = k + max(amount[l][k-1], amount[k+1][h]);
                    }
                    globalmin = min(globalmin, local);
                }
                amount[l][h] = globalmin;
            }
        }
        return amount[1][n];
    }
};

File: MyLeetCode/python/Guess Number Higher or Lower II.py
###
# DP
# Time Complexity: O(n^3)
# Space Complexity: O(n^2)
# State: dp[i][j]: min cost from i to j
# Function: dp[i][j] = min(pivot + max(dp[i][pivot-1], dp[pivot+1][j]) i<=pivot<=[j]
# Initialization: dp[i][i] = 0 i = 1 to n
# Answer: dp
###
class Solution(object):
    def getMoneyAmount(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n <= 1:
            return 0
        dp = [[0] * (n+1) for i in xrange(n+1)]
        for r in xrange(2, n+1):
            for l in xrange(r-1, -1, -1):
                global_min = float('inf')
                for k in xrange(l, r+1):
                    if k == l:
                        local = l + dp[k+1][r]
                    elif k == r:
                        local = r + dp[l][k-1]
                    else:
                        local = k + max(dp[l][k-1], dp[k+1][r])
                    global_min = min(global_min, local)
                dp[l][r] = global_min
        return dp[1][n]
        
        